ibtk/include/ibtk/SnapshotCache.h

```diff
@@ -0,0 +1,226 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2022 - 2022 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBTK_SnapshotCache
+#define included_IBTK_SnapshotCache
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibtk/config.h>
+
+#include <ibtk/ibtk_utilities.h>
+
+#include <tbox/Database.h>
+#include <tbox/Pointer.h>
+#include <tbox/Serializable.h>
+
+#include <CellVariable.h>
+#include <EdgeVariable.h>
+#include <FaceVariable.h>
+#include <HierarchyDataOpsReal.h>
+#include <IntVector.h>
+#include <NodeVariable.h>
+#include <PatchHierarchy.h>
+#include <SideVariable.h>
+
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBTK
+{
+/*!
+ * \brief Class SnapshotCache provides a method of storing snapshots of patch data on a snapshot of the patch hierarchy.
+ *
+ * This class can be used to store one type of variable, which is specified in the constructor.
+ *
+ * When retrieving or storing data with this class, the same patch data layout must be used, e.g. you must consistently
+ * store/retrieve CellData<NDIM, double> with depth of 2.
+ */
+class SnapshotCache : SAMRAI::tbox::Serializable
+{
+public:
+    /*!
+     * The constructor for class SnapshotCache sets some default values, reads in configuration information from input
+     * databases, and registers a variable/context pair with the VariableDatabase. Can optionally set up this class for
+     * restarts.
+     *
+     * The optinal input database is searched for the following keys
+     * - 'refine_type' : String that is the default refine operation for transfering data between hierarchies.
+     * - 'gcw' : Integer array that is the ghost cell width with which the internal variable is set up.
+     *
+     * \note If this object is being created from a restart, the grid geometry used the patch hierarchy must be
+     * supplied. An error will be thrown if the grid geometry is an invalid pointer.
+     */
+    SnapshotCache(std::string object_name,
+                  SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                  SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db = nullptr,
```

Comment:
Similarly, unless there are reasonable defaults we can use 100% of the time, this argument should be mandatory.